File: src/shared/native-backend/capabilities.ts
import type { NativeProtocolVersion } from "./common";

export const NATIVE_BACKEND_NAME = "comando-native-backend";
export const NATIVE_PROTOCOL_VERSION = 1;
export const NATIVE_MINIMUM_CLIENT_PROTOCOL_VERSION = 1;
export const NATIVE_MINIMUM_BACKEND_PROTOCOL_VERSION = 1;

export type NativeCapabilitySet = {
    readonly domains: readonly string[];
    readonly commands: readonly string[];
    readonly events: readonly string[];
    readonly features: readonly string[];
};

export type NativeBackendHandshakeInput = {
    readonly clientName: string;
    readonly clientVersion: string;
    readonly protocolVersion: NativeProtocolVersion;
    readonly supportedProtocolVersions: readonly NativeProtocolVersion[];
};

export type NativeBackendHandshakeOutput = {
    readonly backendName: string;
    readonly backendVersion: string;
    readonly protocolVersion: NativeProtocolVersion;
    readonly minimumClientProtocolVersion: NativeProtocolVersion;
    readonly capabilities: NativeCapabilitySet;
};

export type NativeBackendCapabilitiesOutput = {
    readonly backendName: string;
    readonly backendVersion: string;
    readonly rustVersion: string;
    readonly protocolVersion: NativeProtocolVersion;
    readonly minimumClientProtocolVersion: NativeProtocolVersion;
    readonly minimumBackendProtocolVersion: NativeProtocolVersion;
    readonly capabilities: NativeCapabilitySet;
};

export type NativeProtocolCompatibility = {
    readonly protocolVersion: NativeProtocolVersion;
    readonly minimumClientProtocolVersion: NativeProtocolVersion;
    readonly minimumBackendProtocolVersion: NativeProtocolVersion;
    readonly supportedProtocolVersions: readonly NativeProtocolVersion[];
};

export function parseNativeCapabilitySet(value: unknown): NativeCapabilitySet {
    const record = requireRecord(value, "Native capabilities");

    return {
        domains: parseStringArray(record.domains, "domains"),
        commands: parseStringArray(record.commands, "commands"),
        events: parseStringArray(record.events, "events"),
        features: parseStringArray(record.features, "features"),
    };
}

export function parseNativeBackendHandshakeOutput(
    value: unknown,
): NativeBackendHandshakeOutput {
    const record = requireRecord(value, "Native handshake");

    return {
        backendName: requireString(record.backendName, "backendName"),
        backendVersion: requireString(record.backendVersion, "backendVersion"),
        protocolVersion: requireNumber(record.protocolVersion, "protocolVersion"),
        minimumClientProtocolVersion: requireNumber(
            record.minimumClientProtocolVersion,
            "minimumClientProtocolVersion",
        ),
        capabilities: parseNativeCapabilitySet(record.capabilities),
    };
}

export function parseNativeBackendCapabilitiesOutput(
    value: unknown,
): NativeBackendCapabilitiesOutput {
    const record = requireRecord(value, "Native backend capabilities");

    return {
        backendName: requireString(record.backendName, "backendName"),
        backendVersion: requireString(record.backendVersion, "backendVersion"),
        rustVersion: requireString(record.rustVersion, "rustVersion"),
        protocolVersion: requireNumber(record.protocolVersion, "protocolVersion"),
        minimumClientProtocolVersion: requireNumber(
            record.minimumClientProtocolVersion,
            "minimumClientProtocolVersion",
        ),
        minimumBackendProtocolVersion: requireNumber(
            record.minimumBackendProtocolVersion,
            "minimumBackendProtocolVersion",
        ),
        capabilities: parseNativeCapabilitySet(record.capabilities),
    };
}

export function isNativeProtocolCompatible(input: {
    readonly backendProtocolVersion: NativeProtocolVersion;
    readonly minimumClientProtocolVersion: NativeProtocolVersion;
    readonly clientProtocolVersion?: NativeProtocolVersion;
    readonly minimumBackendProtocolVersion?: NativeProtocolVersion;
}): boolean {
    const clientProtocolVersion =
        input.clientProtocolVersion ?? NATIVE_PROTOCOL_VERSION;
    const minimumBackendProtocolVersion =
        input.minimumBackendProtocolVersion ??
        NATIVE_MINIMUM_BACKEND_PROTOCOL_VERSION;

    return (
        input.backendProtocolVersion >= minimumBackendProtocolVersion &&
        clientProtocolVersion >= input.minimumClientProtocolVersion &&
        input.backendProtocolVersion === NATIVE_PROTOCOL_VERSION
    );
}

function parseStringArray(value: unknown, fieldName: string): readonly string[] {
    if (!Array.isArray(value) || value.some((item) => typeof item !== "string")) {
        throw new Error(`Native capabilities ${fieldName} must be string[].`);
    }

    return value;
}

function requireRecord(
    value: unknown,
    label: string,
): Record<string, unknown> {
    if (typeof value !== "object" || value === null || Array.isArray(value)) {
        throw new Error(`${label} must be an object.`);
    }

    return value as Record<string, unknown>;
}

function requireString(value: unknown, fieldName: string): string {
    if (typeof value !== "string") {
        throw new Error(`Native capability field ${fieldName} must be a string.`);
    }

    return value;
}

function requireNumber(value: unknown, fieldName: string): number {
    if (typeof value !== "number") {
        throw new Error(`Native capability field ${fieldName} must be a number.`);
    }

    return value;
}

File: crates/comando-types/src/capabilities.rs
use serde::{Deserialize, Serialize};

use crate::commands::{BACKEND_COMMANDS, all_commands};
use crate::common::NativeProtocolVersion;
use crate::events::{BACKEND_EVENTS, all_events};
use crate::protocol::NativeRequestMeta;

pub const BACKEND_NAME: &str = "comando-native-backend";
pub const PROTOCOL_VERSION: NativeProtocolVersion = 1;
pub const MINIMUM_CLIENT_PROTOCOL_VERSION: NativeProtocolVersion = 1;
pub const MINIMUM_BACKEND_PROTOCOL_VERSION: NativeProtocolVersion = 1;
pub const RUST_VERSION: &str = "1.96";

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeCapabilitySet {
    pub domains: Vec<String>,
    pub commands: Vec<String>,
    pub events: Vec<String>,
    pub features: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeBackendHandshakeInput {
    pub client_name: String,
    pub client_version: String,
    pub protocol_version: NativeProtocolVersion,
    pub supported_protocol_versions: Vec<NativeProtocolVersion>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeBackendHandshakeOutput {
    pub backend_name: String,
    pub backend_version: String,
    pub protocol_version: NativeProtocolVersion,
    pub minimum_client_protocol_version: NativeProtocolVersion,
    pub capabilities: NativeCapabilitySet,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeBackendCapabilitiesOutput {
    pub backend_name: String,
    pub backend_version: String,
    pub rust_version: String,
    pub protocol_version: NativeProtocolVersion,
    pub minimum_client_protocol_version: NativeProtocolVersion,
    pub minimum_backend_protocol_version: NativeProtocolVersion,
    pub capabilities: NativeCapabilitySet,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeProtocolCompatibility {
    pub protocol_version: NativeProtocolVersion,
    pub minimum_client_protocol_version: NativeProtocolVersion,
    pub minimum_backend_protocol_version: NativeProtocolVersion,
    pub supported_protocol_versions: Vec<NativeProtocolVersion>,
}

pub fn backend_capabilities() -> NativeCapabilitySet {
    NativeCapabilitySet {
        domains: vec![
            "backend".to_string(),
            "persistence".to_string(),
            "projects".to_string(),
            "fs".to_string(),
            "index".to_string(),
            "git".to_string(),
            "terminal".to_string(),
            "settings".to_string(),
            "ai".to_string(),
            "review".to_string(),
            "workspace".to_string(),
        ],
        commands: all_commands().into_iter().map(str::to_string).collect(),
        events: all_events().into_iter().map(str::to_string).collect(),
        features: vec![
            "bootstrap".to_string(),
            "versioned-protocol".to_string(),
            "json-fixtures".to_string(),
        ],
    }
}

pub fn bootstrap_capabilities() -> NativeCapabilitySet {
    NativeCapabilitySet {
        domains: vec!["backend".to_string()],
        commands: BACKEND_COMMANDS
            .iter()
            .map(|command| command.to_string())
            .collect(),
        events: BACKEND_EVENTS
            .iter()
            .map(|event| event.to_string())
            .collect(),
        features: vec!["bootstrap".to_string(), "versioned-protocol".to_string()],
    }
}

pub fn default_request_meta() -> NativeRequestMeta {
    NativeRequestMeta::default()
}

pub fn is_protocol_supported(version: NativeProtocolVersion) -> bool {
    version == PROTOCOL_VERSION
}
